src/platform_impl/macos/event_loop.rs

```diff
@@ -114,6 +164,49 @@ impl<T> EventLoop<T> {
     }
 }
 
+/// Catches panics that happen inside `f` and when a panic
+/// happens, stops the `sharedApplication`
+pub(crate) fn stop_app_on_panic<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Option<R> {
+    match catch_unwind(f) {
+        Ok(r) => Some(r),
+        Err(e) => {
+            // It's important that we set the panic before requesting a `stop`
+            // because some callback are still called during the `stop` message
+            // and we need to know in those callbacks if the application is currently
+            // panicking
+            {
+                let mut panic = CURRENT_PANIC.lock().unwrap();
+                if panic.is_none() {
+                    *panic = Some(e);
+                }
+            }
+            unsafe {
+                let app_class = class!(NSApplication);
+                let app: id = msg_send![app_class, sharedApplication];
+                let _: () = msg_send![app, stop: nil];
+
+                // Posting an dummy event to get stop to take effect immediately.
+                // See: https://stackoverflow.com/questions/48041279/stopping-the-nsapplication-main-event-loop/48064752#48064752
+                let event_class = class!(NSEvent);
+                let dummy_event: id = msg_send![
```

Comment:
It looks like there's code duplication here with the dummy event in `AppState::cleared`, which would be great to consolidate!